### ISSUE TITLE
Elaborate solver attempted to unify int with (Tuple int int) when "using" applied to a pair

### ISSUE THREAD

**Issue by mgritter** [ucsd-progsys/liquidhaskell#2113]
I attempted to use "using" to apply a restriction to the Range type I'd defined, in order to ensure all ranges were valid with start <= end:

```
module Example3 where

type Range = (Int,Int)

{-@ measure start @-}
start :: Range -> Int
start (a,b) = a

{-@ measure end @-}
end :: Range -> Int
end (a,b) = b

{-@ using (Range) as  {r:Range | start r <= end r} @-}

-- seemed to work earlier, now fails
{-@ intsToRanges :: Int -> Int -> Int -> Int -> Maybe (Range,Range) @-}
intsToRanges :: Int -> Int -> Int -> Int -> Maybe (Range,Range)
intsToRanges a b c d = if a <= b && c <= d then Just ((a,b),(c,d)) else Nothing
```

This results in the error:

```
**** LIQUID: ERROR :1:1-1:1: Error
  elaborate solver elabBE 188 "VV##F##14" {VV##F##14 : bool | [(VV##F##14 <=> && [((Example3.start xInv) <= (Example3.end xInv))])]} failed on:
      VV##F##14 <=> Example3.start xInv <= Example3.end xInv
  with error
      The sort (Tuple int int) is not numeric
  because
Cannot unify int with (Tuple int int) in expression: Example3.start xInv 
  in environment
      Example3.end := func(0 , [(Tuple int int); int])

      Example3.start := func(0 , [(Tuple int int); int])

      VV##F##14 := bool

      xInv := (Tuple (Tuple int int) (Tuple int int)) 
```

It looks to me like LH is trying to apply the specification to pairs of Ranges, not just pairs of Ints.

Found using ghc-8.10.1; the online demo also indicates an error but does not show it.

(I worked around the error by defining a data type instead and applied the "using" to that data type.)

**Comment by ranjitjhala** [comment]
@nikivazou  -- after tediously threading `SrcSpan` through the `BindEnv` to make sure the ELAB errors produce source span (PR incoming) I realized this bug is because of `using` on a type alias `Pair` -- I think the code for `using` does not realize that it should _only_ be applied to `(Int, Int)` -- it seems to be applying it to _all_ pairs, including (and here's where the source span is useful) the pair `((a,b),(c,d))` which is causing the problem.

We should either

* ONLY use `using` on datatypes (not aliases) with a specific `data-type constructor` or
* DISABLE it entirely...

Even in the former case, it should have the same restrictions as `measure` -- so if you define 

```haskell
data Pair a b = P a b
``` 

you should not be able to write `using (Pair Int Int) ...` as that is likely to tickle the same *ill-sorted constraint* ...

**Comment by nikivazou** [comment]
Yes, we do the same for the measures. Refine all of them and then prune the unsorted ones. 
I can take a look (it is still open, correct?)

**Comment by ranjitjhala** [comment]
Yup, that would work! (Yes still open, thanks!)
